### ISSUE TITLE
TEST CASE for UIBinding BUG and Validator BUG

### ISSUE THREAD

**Issue by jsfan3** [codenameone/CodenameOne#2472]
In my previous bug reportings (https://github.com/codenameone/CodenameOne/issues/2470 and https://github.com/codenameone/CodenameOne/issues/2471) I didn't provide an easy to check test case.

In this post I provide one test case that demostrates both bugs (on Android 7).

First of all, let's start with two screenshots.

Screenshot 1: I entered the following data. You can note that the Validator doesn't work properly because the "FieldFirstName" is not marked as invalid (it requires at least two chars to be valid):
![screenshot_20180709-054406](https://user-images.githubusercontent.com/1997316/42429939-68e61146-833c-11e8-8cc5-a4806787a238.png)

Screenshot 2: this dialog (opened tapping the "ShowProperties" in the toolbar after entering the data in the previous screenshot) shows that the "firstName" property is "null"!!! Instead that property should contain the text "F" previously entered.

Code to do the test to replicate these bugs:

```
package [...];

import com.codename1.components.SpanLabel;
import static com.codename1.ui.CN.*;
import com.codename1.ui.Form;
import com.codename1.ui.Dialog;
import com.codename1.ui.plaf.UIManager;
import com.codename1.ui.util.Resources;
import com.codename1.io.Log;
import com.codename1.properties.BooleanProperty;
import com.codename1.properties.ListProperty;
import com.codename1.properties.LongProperty;
import com.codename1.properties.Property;
import com.codename1.properties.PropertyBusinessObject;
import com.codename1.properties.PropertyIndex;
import com.codename1.properties.UiBinding;
import com.codename1.ui.Button;
import com.codename1.ui.Toolbar;
import com.codename1.ui.layouts.BoxLayout;
import com.codename1.ui.Container;
import com.codename1.ui.FontImage;
import com.codename1.ui.InputComponent;
import com.codename1.ui.PickerComponent;
import com.codename1.ui.TextArea;
import com.codename1.ui.TextComponent;
import com.codename1.ui.TextField;
import com.codename1.ui.layouts.TextModeLayout;
import com.codename1.ui.validation.Constraint;
import com.codename1.ui.validation.Validator;
import java.util.Calendar;
import java.util.Date;

/**
 * This file was generated by <a href="https://www.codenameone.com/">Codename
 * One</a> for the purpose of building native mobile applications using Java.
 */
public class UIBindingBug {

    private Form current;
    private Resources theme;
    // Data stored in PropertyBusinessObjects
    public static User me = new User();

    public void init(Object context) {

        // Properties registerExternalizable
        me.getPropertyIndex().registerExternalizable();

        // use two network threads instead of one
        updateNetworkThreadCount(2);

        theme = UIManager.initFirstTheme("/theme");

        // Enable Toolbar on all Forms by default
        Toolbar.setGlobalToolbar(true);

        // Pro only feature
        Log.bindCrashProtection(true);

        addNetworkErrorListener(err -> {
            // prevent the event from propagating
            err.consume();
            if (err.getError() != null) {
                Log.e(err.getError());
            }
            Log.sendLogAsync();
            Dialog.show("Connection Error", "There was a networking error in the connection to " + err.getConnectionRequest().getUrl(), "OK", null);
        });

    }

    public void start() {
        if (current != null) {
            current.show();
            return;
        }

        Form inputForm = new Form("Input", BoxLayout.y());
        TextModeLayout tl = new TextModeLayout(7, 1);
        Container userInputCnt = new Container(tl);
        userInputCnt.setUIID("MyContainer01");

        TextComponent firstName = new TextComponent().label("FieldFirstName");
        firstName.getField().setConstraint(TextArea.INITIAL_CAPS_WORD);

        TextComponent familyName = new TextComponent().label("FieldFamilyName");
        familyName.getField().setConstraint(TextArea.INITIAL_CAPS_WORD);

        PickerComponent birthday = PickerComponent.createDate(new Date()).label("PickerBirthday");
        birthday.getPicker().setUseLightweightPopup(true);

        PickerComponent gender = PickerComponent.createStrings(me.genders).label("Gender");
        gender.getPicker().setUseLightweightPopup(true);

        TextComponent email = new TextComponent().label("FieldEmail");
        email.getField().setConstraint(TextField.EMAILADDR);

        TextComponent password = new TextComponent().label("FieldPassword");
        password.getField().setConstraint(TextField.PASSWORD);

        TextComponent pswConfirm = new TextComponent().label("FieldConfirmPassword");
        pswConfirm.getField().setConstraint(TextField.PASSWORD);

        InputComponent.group(firstName, familyName, birthday, gender, email, password, pswConfirm);
        userInputCnt.addAll(firstName, familyName, birthday, gender, email, password, pswConfirm);

        inputForm.add(userInputCnt);
        Button createAccount = new Button("createAccount");
        inputForm.add(createAccount);

        Validator val = new Validator();
        val.addConstraint(firstName, ValidatorUtils.getStringLenghtConstraint(2));
        val.addConstraint(familyName, ValidatorUtils.getStringLenghtConstraint(2));
        val.addConstraint(birthday, ValidatorUtils.get13yearsConstraint());
        val.addConstraint(email, ValidatorUtils.getEmailConstraint());
        val.addConstraint(password, ValidatorUtils.getStringLenghtConstraint(8));
        val.addConstraint(pswConfirm, ValidatorUtils.confirmPassword(password));
        val.addSubmitButtons(createAccount);

        Form hi = new Form("Hi World", BoxLayout.y());
        Button myButton = new Button("Tap here");
        myButton.addActionListener(l -> {
            inputForm.show();
        });
        hi.add(myButton);
        hi.show();

        // Add back command
        inputForm.setBackCommand(
                inputForm.getToolbar().addCommandToLeftBar("Back", FontImage.createMaterial(FontImage.MATERIAL_ARROW_BACK, "Button", 3), ev -> {
                    hi.showBack();
                }));
        
        // Add show Properties command
        inputForm.getToolbar().addCommandToRightBar("ShowProperties", null, ev -> {
                    Dialog.show("PropertyBusinessObject", me.getPropertyIndex().toString(), "OK", "Cancel");
                });

        // Bind the text fields to the User object
        UiBinding uib = new UiBinding();
        uib.bind(me.firstName, firstName.getField());
        uib.bind(me.familyName, familyName.getField());
        uib.bind(me.birthdayDate, birthday.getPicker());
        uib.bind(me.gender, gender.getPicker());
        uib.bind(me.email, email.getField());
        uib.bind(me.password, password.getField());

    }

    public void stop() {
        current = getCurrentForm();
        if (current instanceof Dialog) {
            ((Dialog) current).dispose();
            current = getCurrentForm();
        }
    }

    public void destroy() {
    }

    /**
     * Save the current User object "me" to the Storage
     */
    public static void saveUser() {
        writeObjectToStorage("MyUser", me);
        Log.p("User object written to the Storage");
    }

    /**
     * Restore the User object "me" from the Storage, if it's possible. Returns
     * true if the operation successes, otherwise false (that means that no User
     * object was previously saved).
     *
     * @return true if the User object "me" is restored, false if the operation
     * fails (in that case the value of "me" is not changed)
     */
    public static boolean restoreUser() {
        Object objUser = readObjectFromStorage("MyUser");
        if (objUser != null && objUser instanceof User) {
            me = (User) objUser;
            Log.p("User object restored from the Storage");
            return true;
        } else {
            Log.p("User object cannot be restored from the Storage");
            return false;
        }
    }

}

class User implements PropertyBusinessObject {

    public static final String[] genders = {"Male", "Female", "Other", "Unknown"};

    public final Property<String, User> id = new Property<>("id");
    public final Property<String, User> mainMacroFigure = new Property<>("mainMacroFigure");
    public final Property<String, User> secondMacroFigure = new Property<>("secondMacroFigure");
    public final Property<String, User> personFigure = new Property<>("personFigure");
    public final Property<String, User> organizationFigure = new Property<>("organizationFigure");
    public final Property<String, User> mediaFigure = new Property<>("mediaFigure");

    public final BooleanProperty<User> isIndividual = new BooleanProperty<>("isIndividual");
    public final BooleanProperty<User> isJournalist = new BooleanProperty<>("isJournalist");
    public final BooleanProperty<User> isClubMember = new BooleanProperty<>("isClubMember");

    public final Property<String, User> firstName = new Property<>("firstName");
    public final Property<String, User> familyName = new Property<>("familyName");
    public final Property<String, User> email = new Property<>("email");
    public final Property<String, User> phone = new Property<>("phone");
    public final Property<String, User> gender = new Property<>("gender");
    public final LongProperty<User> birthday
            = new LongProperty<>("birthday");
    public final Property<String, User> avatar = new Property<>("avatar");
    public final Property<String, User> cover = new Property<>("cover");
    public final ListProperty<User, User> friends
            = new ListProperty<>("friends", User.class);
    public final ListProperty<User, User> friendRequests
            = new ListProperty<>("friendRequests", User.class);
    public final ListProperty<User, User> peopleYouMayKnow
            = new ListProperty<>("peopleYouMayKnow", User.class);
    public final Property<String, User> authtoken = new Property<>("authtoken");
    public final Property<String, User> password = new Property<>("password");

    public final Property<Date, User> birthdayDate
            = new Property<Date, User>("birthdayDate", Date.class) {
        @Override
        public Date get() {
            if (birthday.get() == null) {
                return null;
            }
            return new Date(birthday.get());
        }

        @Override
        public User set(Date value) {
            if (value == null) {
                return birthday.set(null);
            }
            return birthday.set(value.getTime());
        }
    };

    private final PropertyIndex idx = new PropertyIndex(this, "User",
            id, mainMacroFigure, secondMacroFigure, personFigure,
            organizationFigure,
            isIndividual, isJournalist, isClubMember,
            firstName, familyName, email, phone, gender, avatar, cover, birthday,
            birthdayDate, friends, friendRequests, peopleYouMayKnow, authtoken,
            password);

    public User() {
        firstName.setLabel("First Name");
        familyName.setLabel("Family Name");
        email.setLabel("FieldEmail");
        phone.setLabel("Phone");
        gender.setLabel("Gender");
        birthdayDate.setLabel("Birthday");
        password.setLabel("FieldPassword");
        idx.setExcludeFromJSON(birthdayDate, true);
        idx.setExcludeFromMap(birthdayDate, true);
    }

    @Override
    public PropertyIndex getPropertyIndex() {
        return idx;
    }

    public String fullName() {
        return firstName.get() + " " + familyName.get();
    }
}

class ValidatorUtils {

    /**
     * String lenght Validator constraint for TextComponent
     *
     * @param stringLenght
     * @return
     */
    public static Constraint getStringLenghtConstraint(int stringLenght) {
        return new Constraint() {

            @Override
            public boolean isValid(Object value) {
                if (value instanceof String) {
                    int currentLenght = ((String) value).length();
                    if (currentLenght >= stringLenght) {
                        return true;
                    }
                }
                return false;
            }

            @Override
            public String getDefaultFailMessage() {
                return "StringTooShort";
            }
        };
    }

    /**
     * (Very simple and minimalistic) Email Validator constraint for
     * TextComponent
     *
     * @param stringLenght
     * @return
     */
    public static Constraint getEmailConstraint() {
        return new Constraint() {

            @Override
            public boolean isValid(Object value) {
                if (value instanceof String) {
                    String email = (String) value;
                    if (email.indexOf('@') > 1
                            && email.indexOf('@') == email.lastIndexOf('@')
                            && email.length() - email.indexOf('@') > 3
                            && email.lastIndexOf('.') > email.indexOf('@')
                            && email.length() - email.lastIndexOf('.') > 2) {
                        return true;
                    }
                }
                return false;
            }

            @Override
            public String getDefaultFailMessage() {
                return "EmailInvalid";
            }
        };
    }

    /**
     * 13-years old Validator constraint for PickerComponent
     *
     * @return 13-years old Validator constraint
     */
    public static Constraint get13yearsConstraint() {
        return new Constraint() {
            @Override
            public boolean isValid(Object value) {
                boolean res = false;
                if (value != null && value instanceof Date) {
                    Calendar birthday = Calendar.getInstance();
                    birthday.setTime((Date) value);
                    Calendar calendar = Calendar.getInstance();
                    calendar.setTime(new Date());
                    calendar.add(Calendar.YEAR, -13);
                    if (birthday.before(calendar)) {
                        res = true;
                    }
                }
                return res;
            }

            @Override
            public String getDefaultFailMessage() {
                return "AreYou13YearsOld";
            }
        };
    }

    /**
     * Confirm password (using the value stored in the User
     * PropertyBusinessObject
     *
     * @param password
     * @return the Validator Constraint
     */
    public static Constraint confirmPassword(TextComponent password) {
        return new Constraint() {
            @Override
            public boolean isValid(Object value) {
                if (value != null && value instanceof String) {
                    String passToConfirm = (String) value;
                    String originalPassword = "";
                    String enteredValue = password.getText();
                    if (enteredValue != null) {
                        originalPassword = enteredValue;
                    }
                    return passToConfirm.equals(originalPassword);
                }
                return false;
            }

            @Override
            public String getDefaultFailMessage() {
                return "PasswordsAreDifferent";
            }
        };
    }

}

```

**Comment by shannah** [comment]
I'm not sure that this is a bug.  me's firstName property gets updated when the firstName field's action event is fired.  The implementation doesn't guarantee that firstName's action event will be fired before ShowProperties' action event is fired.  I'm not aware of any guarantee of the order of these events in the design either.  (Correct me if I'm wrong here @codenameone ).

You can work around this race condition in many different ways.  E.g. you could just delay the Dialog show() slightly.  Simply putting it inside a `callSerially` might do it - or you might need to use a timer for this approach.

Or you could add a property change listener to "me" inside your "ShowProperties" action handler to update the dialog contents when the properties are changed - then remove this listener when the dialog is disposed.

If the binding used DataChangedListeners underneath rather than ActionListeners, this would nearly eliminate the likelihood of a race condition like this, but that's not how it is implemented, and it isn't clear whether the tradeoffs involved in such an implementation would be worth it.

**Comment by jsfan3** [comment]
I thank you for the time you spent for this problem, however did you do the test with an Android 7 real device?

It's a blocking bug the prevents my Android (7?) users to registrate in my app, because it doesn't store in the PropertyBusinessObject the first name, the email address and other data (while it works perfectly on iPhone X and in the Simulator). You wrote: **«me's firstName property gets updated when the firstName field's action event is fired»**, but that's not true on my Huawei BLN-L21 Honor 6x device, because the firstName and the email properties are never updated _(if I understood correctly, in this example I assume that the action event is fired when the user press the "done" or "next" button in the virtual keyboard after writing the first name)_. Some fields (like first name and email) are not updated even when I execute the `writeObjectToStorage("MyUser", me);` inside the ActionListener of a submit button.

This is the log of a more complex app, in which I have the same input Form like in this example. Under the TextModeLayout, I inserted a submit button, which action event listener saves the PropertyBusinessObject (binded to the input fields) to the Storage. In the app I inserted the following data:
first name: Francesco
family name: Galgani
birthday: July 11 1982
gender: male
email: nospam@nospam.com
password: 11111111
confirm password: 11111111

BUT, as you can see in the following log, only the family name, gender and birthday were saved!!!

```
[LOG] 0:0:0,0 - Codename One revisions: ab02d7476d1486e3babf14a0b26a5a5205672439

[LOG] 0:0:0,4 - Native Logs Reader initialized correctly
[LOG] 0:0:0,5 - The following full log contains these sections:
[LOG] 0:0:0,6 - 1. Device tester
[LOG] 0:0:0,7 - 2. Standard CN1 App Log
[LOG] 0:0:0,8 - 3. Native Log (Android Logcat)
[LOG] 0:0:0,9 - 


[LOG] 0:0:0,10 - 


[LOG] 0:0:0,13 - -----------------------
[LOG] 0:0:0,14 - ---- DEVICE TESTER ----
[EDT] 0:0:0,224 - User object cannot be restored from the Storage
[EDT] 0:0:0,225 - User is not logged in
[LOG] 0:0:1,372 - 
------ DEVICE DETECTION -------
Name: honor 6x
Model: BLN-L21
Manufacturer: Huawei
--- END OF DEVICE DETECTION ---


[LOG] 0:0:1,377 - Density: DENSITY_HD
Platform Name: and
User Agent: Dalvik/2.1.0 (Linux; U; Android 7.0; BLN-L21 Build/HONORBLN-L21)
OS: Android
OS Version: 7.0
Display Width X Height: 1080X1740
1mm In Pixels: 18,898
Language: it
Locale: IT
Currency Symbol: €
Are Mutable Images Fast: false
Can Force Orientation: true
Has Camera: true
Badging: false
Desktop: false
Tablet: false
Gaussian Blur Support: true
Get All Contacts Fast: true
Multi Touch: true
PICKER_TYPE_DATE: true
PICKER_TYPE_DATE_AND_TIME: false
PICKER_TYPE_STRINGS: true
PICKER_TYPE_TIME: true
Native Share: true
Native Video Player Controls: true
Notification: true
Open Native Navigation: true
Screen Saver Disable: true
Simulator: false
[LOG] 0:0:1,378 - 


[LOG] 0:0:1,379 - -----------------------
[LOG] 0:0:1,379 - ------- APP LOG -------
[EDT] 0:0:5,643 - mypackage.registration.c.g Form shown correctly
[EDT] 0:0:6,630 - The user goes forward from the Termini legali Form to the Seleziona un gruppo Form
[EDT] 0:0:6,775 - mypackage.registration.c.i Form shown correctly
[EDT] 0:0:8,927 - mypackage.registration.c.ac Form shown correctly
[EDT] 0:1:54,455 - User object written to the Storage, with the following data:
User : {
id = null
mainMacroFigure = MacroFigureBusiness
secondMacroFigure = null
personFigure = null
organizationFigure = null
isIndividual = null
isJournalist = null
isClubMember = null
firstName = null
familyName = Galgani
email = null
phone = null
gender = Maschio
avatar = null
cover = null
birthday = 395269943535
birthdayDate = null
friends = friends = null
friendRequests = friendRequests = null
peopleYouMayKnow = peopleYouMayKnow = null
authtoken = null
password = null
}
[EDT] 0:2:34,823 - mypackage.registration.c.ac Form shown correctly
[EDT] 0:2:47,433 - mypackage.registration.c.ag Form shown correctly
[EDT] 0:2:50,122 - 


[EDT] 0:2:50,124 - ------------------------
[EDT] 0:2:50,125 - ---- ANDROID LOGCAT ----
[EDT] 0:2:50,126 - --------- beginning of main
07-11 23:11:42.663  9065  9065 I hwaps   : JNI_OnLoad
07-11 23:11:42.774  9065  9070 I art     : Do partial code cache collection, code=98KB, data=124KB
07-11 23:11:42.775  9065  9070 I art     : After code cache collection, code=94KB, data=122KB
07-11 23:11:42.775  9065  9070 I art     : Increasing code cache capacity to 512KB
07-11 23:11:42.789  9065  9091 E OpenGLRenderer: allen debug liyu Key: 68719476736
07-11 23:11:43.863  9065  9086 I System.out: showKeyboard false
07-11 23:11:43.890  9065  9091 E OpenGLRenderer: allen debug liyu Key: 103079215104
07-11 23:11:43.990  9065  9070 I art     : Compiler allocated 4MB to compile void com.codename1.p.g.a.a(com.codename1.p.v, int, int, int, int, com.codename1.p.l)
07-11 23:11:44.265  9065  9086 E HW-JPEG-DEC: [HME_JPEG_DEC_Delete](3321): HME_JPEG_DEC_Delete: decoder_ctx=null
07-11 23:11:44.821  9065  9091 E OpenGLRenderer: allen debug liyu Key: 34359738369
07-11 23:11:45.046  9065  9070 I art     : Do full code cache collection, code=210KB, data=251KB
07-11 23:11:45.046  9065  9070 I art     : After code cache collection, code=184KB, data=215KB
07-11 23:11:45.061  9065  9086 E HW-JPEG-DEC: [HME_JPEG_DEC_Delete](3321): HME_JPEG_DEC_Delete: decoder_ctx=null
07-11 23:11:45.144  9065  9086 E HW-JPEG-DEC: [HME_JPEG_DEC_Delete](3321): HME_JPEG_DEC_Delete: decoder_ctx=null
07-11 23:11:45.168  9065  9086 E HW-JPEG-DEC: [HME_JPEG_DEC_Delete](3321): HME_JPEG_DEC_Delete: decoder_ctx=null
07-11 23:11:45.190  9065  9086 E HW-JPEG-DEC: [HME_JPEG_DEC_Delete](3321): HME_JPEG_DEC_Delete: decoder_ctx=null
07-11 23:11:45.215  9065  9086 E HW-JPEG-DEC: [HME_JPEG_DEC_Delete](3321): HME_JPEG_DEC_Delete: decoder_ctx=null
07-11 23:11:45.238  9065  9086 I System.out: showKeyboard false
07-11 23:11:45.252  9065  9086 E HW-JPEG-DEC: [HME_JPEG_DEC_Delete](3321): HME_JPEG_DEC_Delete: decoder_ctx=null
07-11 23:11:45.255  9065  9086 E HW-JPEG-DEC: [HME_JPEG_DEC_Delete](3321): HME_JPEG_DEC_Delete: decoder_ctx=null
07-11 23:11:45.255  9065  9086 E HW-JPEG-DEC: [HME_JPEG_DEC_Delete](3321): HME_JPEG_DEC_Delete: decoder_ctx=null
07-11 23:11:45.256  9065  9086 E HW-JPEG-DEC: [HME_JPEG_DEC_Delete](3321): HME_JPEG_DEC_Delete: decoder_ctx=null
07-11 23:11:45.256  9065  9086 E HW-JPEG-DEC: [HME_JPEG_DEC_Delete](3321): HME_JPEG_DEC_Delete: decoder_ctx=null
07-11 23:11:45.257  9065  9086 E HW-JPEG-DEC: [HME_JPEG_DEC_Delete](3321): HME_JPEG_DEC_Delete: decoder_ctx=null
07-11 23:11:45.257  9065  9086 E HW-JPEG-DEC: [HME_JPEG_DEC_Delete](3321): HME_JPEG_DEC_Delete: decoder_ctx=null
07-11 23:11:45.281  9065  9091 E OpenGLRenderer: allen debug liyu Key: 4
07-11 23:11:46.256  9065  9086 I System.out: showKeyboard false
07-11 23:11:46.342  9065  9070 I art     : Do partial code cache collection, code=196KB, data=232KB
07-11 23:11:46.343  9065  9070 I art     : After code cache collection, code=151KB, data=183KB
07-11 23:11:46.343  9065  9070 I art     : Increasing code cache capacity to 1024KB
07-11 23:11:47.152  9065  9091 E OpenGLRenderer: allen debug liyu Key: 34359738368
07-11 23:11:47.351  9065  9086 E HW-JPEG-DEC: [HME_JPEG_DEC_Delete](3321): HME_JPEG_DEC_Delete: decoder_ctx=null
07-11 23:11:48.502  9065  9086 I System.out: showKeyboard false
07-11 23:11:49.993  9065  9065 E HW-JPEG-DEC: [HME_JPEG_DEC_Delete](3321): HME_JPEG_DEC_Delete: decoder_ctx=null
07-11 23:11:49.995  9065  9065 E HW-JPEG-DEC: [HME_JPEG_DEC_Delete](3321): HME_JPEG_DEC_Delete: decoder_ctx=null
07-11 23:11:49.997  9065  9065 E HW-JPEG-DEC: [HME_JPEG_DEC_Delete](3321): HME_JPEG_DEC_Delete: decoder_ctx=null
07-11 23:11:50.014  9065  9065 I InPlaceEditView: showVirtualKeyboard show=true
07-11 23:11:50.016  9065  9065 I HwSecImmHelper: mSecurityInputMethodService is null
07-11 23:11:50.097  9065  9077 I InPlaceEditView.ResultReceiver: resultCode = RESULT_SHOWN
07-11 23:11:54.249  9065  9065 E HW-JPEG-DEC: [HME_JPEG_DEC_Delete](3321): HME_JPEG_DEC_Delete: decoder_ctx=null
07-11 23:11:54.250  9065  9065 E HW-JPEG-DEC: [HME_JPEG_DEC_Delete](3321): HME_JPEG_DEC_Delete: decoder_ctx=null
07-11 23:11:54.251  9065  9065 E HW-JPEG-DEC: [HME_JPEG_DEC_Delete](3321): HME_JPEG_DEC_Delete: decoder_ctx=null
07-11 23:11:54.276  9065  9065 I InPlaceEditView: showVirtualKeyboard show=true
07-11 23:11:54.278  9065  9065 I HwSecImmHelper: mSecurityInputMethodService is null
07-11 23:11:54.290  9065  9065 W IInputConnectionWrapper: getExtractedText on inactive InputConnection
07-11 23:11:54.298  9065  9065 W IInputConnectionWrapper: getTextBeforeCursor on inactive InputConnection
07-11 23:11:54.315  9065  9065 W IInputConnectionWrapper: getTextAfterCursor on inactive InputConnection
07-11 23:11:54.335  9065  9065 W IInputConnectionWrapper: getSelectedText on inactive InputConnection
07-11 23:11:54.349  9065  9065 W IInputConnectionWrapper: getExtractedText on inactive InputConnection
07-11 23:11:54.366  9065  9065 W IInputConnectionWrapper: getTextBeforeCursor on inactive InputConnection
07-11 23:11:54.383  9065  9065 W IInputConnectionWrapper: getTextAfterCursor on inactive InputConnection
07-11 23:11:54.400  9065  9065 W IInputConnectionWrapper: getSelectedText on inactive InputConnection
07-11 23:11:54.401  9065  9065 W IInputConnectionWrapper: getTextBeforeCursor on inactive InputConnection
07-11 23:11:54.401  9065  9065 W IInputConnectionWrapper: endBatchEdit on inactive InputConnection
07-11 23:11:54.451  9065  9077 I InPlaceEditView.ResultReceiver: resultCode = RESULT_SHOWN
07-11 23:11:58.562  9065  9065 I HwSecImmHelper: mSecurityInputMethodService is null
07-11 23:11:58.563  9065  9065 I System.out: -------In endEditing
07-11 23:11:58.563  9065  9065 I System.out: Next flag: true
07-11 23:11:58.564  9065  9065 I System.out: Next cmp: k[x=48 y=97 width=984 height=119 name=null, text = ..., gap = 2]
07-11 23:11:58.564  9065  9065 I System.out: Reason : 1 IME Action: 1
07-11 23:11:58.565  9065  9065 I System.out: -----LEAVE KEYBOARD SHOWING: false
07-11 23:11:58.565  9065  9065 I System.out: Hiding virtual keyboard
07-11 23:11:58.565  9065  9065 I InPlaceEditView: showVirtualKeyboard show=false
07-11 23:11:58.567  9065  9065 I HwSecImmHelper: mSecurityInputMethodService is null
07-11 23:11:58.571  9065  9065 W IInputConnectionWrapper: getExtractedText on inactive InputConnection
07-11 23:11:58.573  9065  9065 W IInputConnectionWrapper: getTextBeforeCursor on inactive InputConnection
07-11 23:11:58.574  9065  9065 W IInputConnectionWrapper: getTextAfterCursor on inactive InputConnection
07-11 23:11:58.575  9065  9065 W IInputConnectionWrapper: getSelectedText on inactive InputConnection
07-11 23:11:58.576  9065  9065 W IInputConnectionWrapper: endBatchEdit on inactive InputConnection
07-11 23:11:58.588  9065  9065 W IInputConnectionWrapper: beginBatchEdit on inactive InputConnection
07-11 23:11:58.588  9065  9065 W IInputConnectionWrapper: finishComposingText on inactive InputConnection
07-11 23:11:58.605  9065  9065 W IInputConnectionWrapper: endBatchEdit on inactive InputConnection
07-11 23:11:58.677  9065  9086 I System.out: showKeyboard false
07-11 23:11:58.710  9065  9076 I InPlaceEditView.ResultReceiver: resultCode = RESULT_HIDDEN
07-11 23:11:59.474  9065  9070 I art     : Compiler allocated 4MB to compile void com.codename1.p.g.a.a(com.codename1.p.v, int, int, int, int, com.codename1.p.l)
07-11 23:12:12.564  9065  9086 E HW-JPEG-DEC: [HME_JPEG_DEC_Delete](3321): HME_JPEG_DEC_Delete: decoder_ctx=null
07-11 23:12:17.855  9065  9086 E HW-JPEG-DEC: [HME_JPEG_DEC_Delete](3321): HME_JPEG_DEC_Delete: decoder_ctx=null
07-11 23:12:35.516  9065  9065 E HW-JPEG-DEC: [HME_JPEG_DEC_Delete](3321): HME_JPEG_DEC_Delete: decoder_ctx=null
07-11 23:12:35.517  9065  9065 E HW-JPEG-DEC: [HME_JPEG_DEC_Delete](3321): HME_JPEG_DEC_Delete: decoder_ctx=null
07-11 23:12:35.519  9065  9065 E HW-JPEG-DEC: [HME_JPEG_DEC_Delete](3321): HME_JPEG_DEC_Delete: decoder_ctx=null
07-11 23:12:35.530  9065  9065 I InPlaceEditView: showVirtualKeyboard show=true
07-11 23:12:35.532  9065  9065 I HwSecImmHelper: mSecurityInputMethodService is null
07-11 23:12:35.603  9065  9076 I InPlaceEditView.ResultReceiver: resultCode = RESULT_SHOWN
07-11 23:13:01.547  9065  9065 I InPlaceEditView: showVirtualKeyboard show=true
07-11 23:13:01.548  9065  9065 I HwSecImmHelper: mSecurityInputMethodService is null
07-11 23:13:01.560  9065  9065 W IInputConnectionWrapper: getExtractedText on inactive InputConnection
07-11 23:13:01.571  9065  9065 W IInputConnectionWrapper: getTextBeforeCursor on inactive InputConnection
07-11 23:13:01.588  9065  9065 W IInputConnectionWrapper: getTextAfterCursor on inactive InputConnection
07-11 23:13:01.605  9065  9065 W IInputConnectionWrapper: getSelectedText on inactive InputConnection
07-11 23:13:01.623  9065  9065 W IInputConnectionWrapper: endBatchEdit on inactive InputConnection
07-11 23:13:01.694  9065  9076 I InPlaceEditView.ResultReceiver: resultCode = RESULT_SHOWN
07-11 23:13:26.454  9065  9065 I InPlaceEditView: showVirtualKeyboard show=true
07-11 23:13:26.456  9065  9065 I HwSecImmHelper: mSecurityInputMethodService is null
07-11 23:13:26.470  9065  9065 W IInputConnectionWrapper: endBatchEdit on inactive InputConnection
07-11 23:13:26.470  9065  9065 W IInputConnectionWrapper: beginBatchEdit on inactive InputConnection
07-11 23:13:26.470  9065  9065 W IInputConnectionWrapper: finishComposingText on inactive InputConnection
07-11 23:13:26.470  9065  9065 W IInputConnectionWrapper: endBatchEdit on inactive InputConnection
07-11 23:13:26.470  9065  9065 W IInputConnectionWrapper: beginBatchEdit on inactive InputConnection
07-11 23:13:26.470  9065  9065 W IInputConnectionWrapper: finishComposingText on inactive InputConnection
07-11 23:13:26.470  9065  9065 W IInputConnectionWrapper: endBatchEdit on inactive InputConnection
07-11 23:13:26.496  9065  9065 W IInputConnectionWrapper: beginBatchEdit on inactive InputConnection
07-11 23:13:26.496  9065  9065 W IInputConnectionWrapper: finishComposingText on inactive InputConnection
07-11 23:13:26.496  9065  9065 W IInputConnectionWrapper: endBatchEdit on inactive InputConnection
07-11 23:13:26.496  9065  9065 W IInputConnectionWrapper: beginBatchEdit on inactive InputConnection
07-11 23:13:26.496  9065  9065 W IInputConnectionWrapper: finishComposingText on inactive InputConnection
07-11 23:13:26.496  9065  9065 W IInputConnectionWrapper: endBatchEdit on inactive InputConnection
07-11 23:13:26.540  9065  9076 I InPlaceEditView.ResultReceiver: resultCode = RESULT_SHOWN
07-11 23:13:32.082  9065  9065 I HwSecImmHelper: mSecurityInputMethodService is null
07-11 23:13:32.086  9065  9065 I System.out: -------In endEditing
07-11 23:13:32.086  9065  9065 I System.out: Next flag: false
07-11 23:13:32.087  9065  9065 I System.out: Next cmp: null
07-11 23:13:32.087  9065  9065 I System.out: Reason : 1 IME Action: 1
07-11 23:13:32.087  9065  9065 I System.out: -----LEAVE KEYBOARD SHOWING: true
07-11 23:13:32.087  9065  9065 I System.out: Hiding virtual keyboard
07-11 23:13:32.087  9065  9065 I InPlaceEditView: showVirtualKeyboard show=false
07-11 23:13:32.088  9065  9065 I HwSecImmHelper: mSecurityInputMethodService is null
07-11 23:13:32.108  9065  9065 W IInputConnectionWrapper: beginBatchEdit on inactive InputConnection
07-11 23:13:32.108  9065  9065 W IInputConnectionWrapper: finishComposingText on inactive InputConnection
07-11 23:13:32.108  9065  9065 W IInputConnectionWrapper: endBatchEdit on inactive InputConnection
07-11 23:13:32.130  9065  9065 W IInputConnectionWrapper: beginBatchEdit on inactive InputConnection
07-11 23:13:32.130  9065  9065 W IInputConnectionWrapper: finishComposingText on inactive InputConnection
07-11 23:13:32.130  9065  9065 W IInputConnectionWrapper: endBatchEdit on inactive InputConnection
07-11 23:13:32.130  9065  9065 W IInputConnectionWrapper: beginBatchEdit on inactive InputConnection
07-11 23:13:32.130  9065  9065 W IInputConnectionWrapper: finishComposingText on inactive InputConnection
07-11 23:13:32.130  9065  9065 W IInputConnectionWrapper: endBatchEdit on inactive InputConnection
07-11 23:13:32.130  9065  9065 W IInputConnectionWrapper: beginBatchEdit on inactive InputConnection
07-11 23:13:32.130  9065  9065 W IInputConnectionWrapper: finishComposingText on inactive InputConnection
07-11 23:13:32.130  9065  9065 W IInputConnectionWrapper: endBatchEdit on inactive InputConnection
07-11 23:13:33.982  9065  9086 I System.out: showKeyboard false
07-11 23:14:14.354  9065  9065 I System.out: -------In endEditing
07-11 23:14:14.427  9065  9086 I System.out: showKeyboard false
07-11 23:14:15.608  9065  9065 E HW-JPEG-DEC: [HME_JPEG_DEC_Delete](3321): HME_JPEG_DEC_Delete: decoder_ctx=null
07-11 23:14:15.609  9065  9065 E HW-JPEG-DEC: [HME_JPEG_DEC_Delete](3321): HME_JPEG_DEC_Delete: decoder_ctx=null
07-11 23:14:15.610  9065  9065 E HW-JPEG-DEC: [HME_JPEG_DEC_Delete](3321): HME_JPEG_DEC_Delete: decoder_ctx=null
07-11 23:14:15.629  9065  9065 I InPlaceEditView: showVirtualKeyboard show=true
07-11 23:14:15.631  9065  9065 I HwSecImmHelper: mSecurityInputMethodService is null
07-11 23:14:15.712  9065  9327 I InPlaceEditView.ResultReceiver: resultCode = RESULT_SHOWN
07-11 23:14:15.776  9065  9070 I art     : Do full code cache collection, code=466KB, data=502KB
07-11 23:14:15.778  9065  9070 I art     : After code cache collection, code=348KB, data=374KB
07-11 23:14:19.863  9065  9065 I System.out: -------In endEditing
07-11 23:14:19.863  9065  9065 I System.out: Next flag: true
07-11 23:14:19.864  9065  9065 I System.out: Next cmp: an$1[x=48 y=97 width=984 height=121 name=null]
07-11 23:14:19.864  9065  9065 I System.out: Reason : 3 IME Action: 1
07-11 23:14:19.864  9065  9065 I System.out: -----LEAVE KEYBOARD SHOWING: false
07-11 23:14:19.864  9065  9065 I System.out: Hiding virtual keyboard
07-11 23:14:19.864  9065  9065 I InPlaceEditView: showVirtualKeyboard show=false
07-11 23:14:19.865  9065  9065 I HwSecImmHelper: mSecurityInputMethodService is null
07-11 23:14:19.871  9065  9327 I InPlaceEditView.ResultReceiver: resultCode = RESULT_HIDDEN
07-11 23:14:19.876  9065  9065 W IInputConnectionWrapper: beginBatchEdit on inactive InputConnection
07-11 23:14:19.876  9065  9065 W IInputConnectionWrapper: finishComposingText on inactive InputConnection
07-11 23:14:19.876  9065  9065 W IInputConnectionWrapper: endBatchEdit on inactive InputConnection
07-11 23:14:19.876  9065  9065 W IInputConnectionWrapper: beginBatchEdit on inactive InputConnection
07-11 23:14:19.876  9065  9065 W IInputConnectionWrapper: finishComposingText on inactive InputConnection
07-11 23:14:19.876  9065  9065 W IInputConnectionWrapper: endBatchEdit on inactive InputConnection
07-11 23:14:19.916  9065  9065 W IInputConnectionWrapper: beginBatchEdit on inactive InputConnection
07-11 23:14:19.916  9065  9065 W IInputConnectionWrapper: finishComposingText on inactive InputConnection
07-11 23:14:19.916  9065  9065 W IInputConnectionWrapper: endBatchEdit on inactive InputConnection
07-11 23:14:19.916  9065  9065 W IInputConnectionWrapper: beginBatchEdit on inactive InputConnection
07-11 23:14:19.916  9065  9065 W IInputConnectionWrapper: finishComposingText on inactive InputConnection
07-11 23:14:19.916  9065  9065 W IInputConnectionWrapper: endBatchEdit on inactive InputConnection
07-11 23:14:21.231  9065  9086 I System.out: showKeyboard false
07-11 23:14:25.432  9065  9086 I System.out: showKeyboard false
07-11 23:14:26.473  9065  9086 I System.out: [logcat, -d]
07-11 23:14:26.473  9065  9086 I System.out: null
07-11 23:14:26.473  9065  9086 I System.out: null
07-11 23:14:26.473  9065  9086 I System.out: Calling by::className:a.a.a.a.c  MethodName:b
```

To compare, this is the log of the same app on iPhone with the same input data:

```
[LOG] 0:0:0,0 - Codename One revisions: 71abfb0f6d3fce763c5c11d6b7101925e25d8335

[LOG] 0:0:0,7 - Native Logs Reader initialized correctly
[LOG] 0:0:0,8 - The following full log contains these sections:
[LOG] 0:0:0,8 - 1. Device tester
[LOG] 0:0:0,8 - 2. Standard CN1 App Log
[LOG] 0:0:0,8 - 


[LOG] 0:0:0,9 - 


[LOG] 0:0:0,9 - -----------------------
[LOG] 0:0:0,9 - ---- DEVICE TESTER ----
[EDT] 0:0:0,30 - User object cannot be restored from the Storage
[EDT] 0:0:0,30 - User is not logged in
[LOG] 0:0:0,461 - 
------ DEVICE DETECTION -------
Name: iPhone X 
Model: iPhone X (Global)
Manufacturer: Apple Inc.
--- END OF DEVICE DETECTION ---


[LOG] 0:0:0,506 - Density: DENSITY_560
Platform Name: ios
User Agent: Mozilla/5.0 (iPhone; CPU iPhone OS 11_4 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Mobile/15F79
OS: iOS
OS Version: 11.4
Display Width X Height: 1125X2436
1mm In Pixels: 18,031
Language: it
Locale: IT
Currency Symbol: EUR
Are Mutable Images Fast: false
Can Force Orientation: true
Has Camera: true
Badging: true
Desktop: false
Tablet: false
Gaussian Blur Support: true
Get All Contacts Fast: false
Multi Touch: true
PICKER_TYPE_DATE: true
PICKER_TYPE_DATE_AND_TIME: true
PICKER_TYPE_STRINGS: true
PICKER_TYPE_TIME: true
Native Share: true
Native Video Player Controls: true
Notification: false
Open Native Navigation: true
Screen Saver Disable: true
Simulator: false
[LOG] 0:0:0,507 - 


[LOG] 0:0:0,507 - -----------------------
[LOG] 0:0:0,507 - ------- APP LOG -------
[EDT] 0:0:2,923 - mypackage.registration.forms.Form03LegalTerms Form shown correctly
[EDT] 0:0:3,876 - The user goes forward from the Termini legali Form to the Seleziona un gruppo Form
[EDT] 0:0:3,914 - mypackage.registration.forms.Form04ASelectGroup Form shown correctly
[EDT] 0:0:6,121 - mypackage.registration.forms.Form07Registry Form shown correctly
[EDT] 0:1:26,624 - User object written to the Storage, with the following data:
User : {
id = null
mainMacroFigure = MacroFigureBusiness
secondMacroFigure = null
personFigure = null
organizationFigure = null
isIndividual = null
isJournalist = null
isClubMember = null
firstName = Francesco
familyName = Galgani
email = nospam@nospam.com
phone = null
gender = Maschio
avatar = null
cover = null
birthday = 395270914866
birthdayDate = null
friends = friends = null
friendRequests = friendRequests = null
peopleYouMayKnow = peopleYouMayKnow = null
authtoken = null
password = 11111111
}
[EDT] 0:1:59,215 - mypackage.registration.forms.ShowLog Form shown correctly
[EDT] 0:2:3,561 - 
```

**Comment by jsfan3** [comment]
Moreover, there is also the problem of the Validator (on Android), that works only for "family name", "birthday" and "confirm password".

"First name", "Email" and "Password" have a validator, but it doesn't work (the input is always validates, even if it's blank)... and "First name", "Email" and "Password" are not binded to the BusinessPropertyObject... so I guess that the two bugs are correlated.

Another info: these bug, on my Android device and using the this test case, always replicates.

**Comment by codenameone** [comment]
Maybe I'm missing something here but I'm not sure what Dialog we are talking about. All the dialogs I see in the code are not a part of the input flow. The order of the events shouldn't matter to the validator or binding either. This might be a bug there but I didn't have time to go into the code right now with everything that's going on.

**Comment by shannah** [comment]
The click on the "ShowProperties" button, will trigger "unfocus/end editing" on the text field, and it will also trigger the action listeners of the ShowProperties button.  It is in the action event of ShowProperties that the modal dialog is shown where this test case demonstrates that the input of the text field hasn't propagated to its binding yet.

Although the unfocus/end editing trigger on the text field will always happen before the show properties button action event (in the case of a button click), that doesn't necessarily mean that text field's action event will be triggered before the button's action listener.  This is because the unfocus has to do async cross-thread communication to the native thread to end the native text editing, which will ultimately trigger an action event on the EDT.  That final action event may occur after the button's action event.

**Comment by codenameone** [comment]
In that case even a callSerially might fail as the native update might skip. I see. If you do the input and then focus back does it "fix that"?

In that case maybe we should implement the binding on data change instead of action event?

**Comment by shannah** [comment]
Yes.  Call serially might not be enough either.  A timer might be necessary to delay it.

Using DataChangeEvents would be more consistent with other binding solutions I've seen (e.g. angular) - though would be noisier and would only work on TextField objects - not TextArea in general.

Another possibility would be to supplement with focus listeners.

**Comment by codenameone** [comment]
Looking at this again. The dialog is launched on a button press. Won't that trigger a focus lost or text field commit?
If not, I think it should and at least used to. Even if the field is still editable I think clicking outside it should trigger an action event. Does that make sense?

**Comment by shannah** [comment]
Long response here.  Short version is that I recommend adding a "change" event to TextArea that would be fired by both "focus lost" and "editing complete", but only if the text content has changed since the last change event was fired.  Then we use these events in the Binding code.

Long response:

> Even if the field is still editable I think clicking outside it should trigger an action event. Does that make sense?

Makes sense.  Just took a look at the code.  The only things that trigger action events are:
1. Designated key is typed into field (if key is one of the designated 'end editing' keys).  (TextField only)
2. Pointer clicked in textfield (text field only)
3. On Editing Complete event (textarea only).

Adding an additional call when the user presses outside the field but there are a few sharp edges to deal with.  Do we have a definition of how it is supposed to work?  We are already deviating from Swing's JTextField action event behaviour - they only fire when user hits "Enter".  Not even a tab out will trigger an action event on their text fields.  

If we fire an action event also on focus lost (or on a pointer event outside the field), it should fix this issue.  But it will result in 2 action events being sent in that case.  1 on focus lost, and another when editing is complete.  That might not be a big deal - not sure.  It might trip up apps that were written with the assumption that action events are fired when editing is complete only (even though cases 1 and 2 listed above will already result in 2 action events on editing complete).   I could try to deal with that case by recording if the event has already been sent, and *not* send it again on editing complete if it has already been triggered by a pointer event.  That still presents a possibility of regressions as the order of events will have been changed.

Sorry for being so wordy here... seems like a simple thing but it is kind of murky because of the loose definition (or lack of definition) of what an action event on a TextArea means.  Some people use it as a "change" event to keep track of changes to the field.  Some use it as an "on editing complete" event to be notified when the user finishes editing the field.  In the first case, they don't care if it is a little noisy.  In the second case, they may be unhappy if it all of a sudden gets noisier.  E.g. if they are modifying the UI inside the action listener, they all of a sudden start making 2 modifications, popping up 2 dialogs, etc.. in their action listener.   And the order may matter too in the latter case -- simply firing the event earlier might break things.

Possible solution:  Add the following to TextArea
 - addChangeListener(ActionEvent)
 - removeChangeListener(ActionEvent)

Change events would be fired whenever the text is changed.  It would be fired on focus lost and editing complete - but only if the text has changed since the last change event.  That way we can be as noisy as we want here.  It would be easy to implement - it is easy to understand - and because it just uses action listeners, it would be easy as a drop-in replacement for action listeners in the binding logic.

**Comment by codenameone** [comment]
I'm for making that fix and guarding against some of the duplicate events. 
Action event was always a bit noisy on text field if someone is relying on it firing only once that's a bug that they should fix. Having two listeners that do "almost the same thing" seems problematic in terms of the API.

Having said that I'm not averse to moving the datachange listener up the chain to text area. That would also let us solve the `UIBinding` issue and make things consistent. 

I had no idea Swing didn't fire on tab out. Sounds like a mistake in Swing IMHO ;-)

**Comment by jsfan3** [comment]
Thank you. When will the Codename One project libs and build servers be updated, so that I can test the changes?

**Comment by codenameone** [comment]
It should have been up for a while. If it still happens then the fix failed.

**Comment by jsfan3** [comment]
In my Netbeans project "UIBindingBug" (which code is reported at the top of this thread), I restored the original libraries for compiling (CLDC11.jar, CodenameOne.jar and CodenameOne_SRC.zip) and for running (JavaSE.jar), then I executed "Update Project Libs" (but no update was downloaded), finally I sent a new Android build. I tested that build on my Android 7 phone...

I entered this information:
FieldFirstName: Francesco
FieldFamilyName: Galgani
PickerBirthday: July 14 2000
Gender: Male
FieldEmail: nospam@nospam.com
FieldPassword: 11111111
FieldConfirmPassword: 11111111

Then I tapped on "Create account" (that does nothing, I tapped it only to lost the focus on the TextModeLayout), finally I tapped "ShowProperties". Same result, the fix is failed:

fistName = null
familyName = Galgani
email = null
gender = Male
birthday = 963571604984
password = null

I reproduced all these steps yesterday and few minutes ago.

**Comment by codenameone** [comment]
Reproduced on a new device build right?

**Comment by jsfan3** [comment]
Yes, of course, I tested it with a new device build.

**Comment by jsfan3** [comment]
At 3.30.33 UTC of today (Fri Jul 20 2018) the UIBinding Bug is still present. Two questions:

- Is your fix already in the build server or will it be available in the next hours?
- This discussion is technical complex for me, I understood only that you focused on a race condition... but, if I understood correctly, a race condition produces a random behaviour, instead in this case I gave you a code that **always** produces the same behavior: have you experienced the issue in the same way I described it?

Thanks.

**Comment by jsfan3** [comment]
I used as compiling library in my UIBindingBug Netbeans project my local fork of Codename One that, of course, I synchronized with the current version, including all your commits.

**Comment by codenameone** [comment]
We pushed a fix on Thursday. We're still thinking about pushing another version over the weekend.
@shannah do you think this (or a different fix) should go in?

**Comment by shannah** [comment]
If you plan to redeploy, this fix is worth including.

**Comment by jsfan3** [comment]
_(In the while, can you reopen this issue?)_

**Comment by shannah** [comment]
I'd prefer to keep it closed.  If, after the server is deployed, you find that you are still experiencing the problem, you can request to have it opened.

**Comment by jsfan3** [comment]
Ok, Can you give me an alert (even commenting here) when I can test again?

**Comment by codenameone** [comment]
FYI forgot to comment tomorrow but we pushed out an update around noon yesterday

**Comment by jsfan3** [comment]
Tested now sending a new build. Now the UIBinding issue seems resolved, but... about this code, there are still two bug related to the Validator. Now I'm going to open a new issue thread.
